### PR TITLE
Doxygen documentation for class refcount_base, refcount, and ref

### PR DESCRIPTION
**Pull request by killerswin2**
This is my first step into learning more about the intercept code base. All changes are related to documentation until I understand more. Wherever possible I tried to link to other docs, or add a url on the topic.

### CODE REVIEW

File: src/client/headers/shared/containers.hpp [hunk 1]
```diff
@@ -200,29 +200,58 @@ namespace intercept::types {
...
+     */
     class refcount_base {
     public:
         constexpr refcount_base() noexcept : _refcount(0) {}
         constexpr refcount_base(const refcount_base&) noexcept : _refcount(0) {}
         constexpr void operator=(const refcount_base&) const noexcept {}
 
+        /**
+         * @brief Increaments refernce counter
+         *
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:216]:
```suggestion
         * @brief Increments the reference counter
```

File: src/client/headers/shared/containers.hpp [hunk 2]
```diff
@@ -200,29 +200,58 @@ namespace intercept::types {
...
+        /**
+         * @brief Decreaments refernce counter
+         * 
+         */
         constexpr int dec_ref() const noexcept {
             return --_refcount;
         }
+        /**
+         * @brief Returns refernce count
+         *
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:230]:
```suggestion
         * @brief Returns the reference count
```

File: src/client/headers/shared/containers.hpp [hunk 3]
```diff
@@ -200,29 +200,58 @@ namespace intercept::types {
...
+     * refcount has to be the first element in a class. Use _refcount_vtable_dummy instead if refcount is preceded by a vtable
+     * <a href="https://en.wikipedia.org/wiki/Reference_counting"> Reference counting </a>
+     */
     class refcount : public refcount_base {
     public:
         virtual ~refcount() = default;
+
+        /**
+         * @brief decrement reference and delete object if refcount is zero
+         *
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:253]:
```suggestion
         * @brief decrement reference count and delete object if refcount is zero
```

File: src/client/headers/shared/containers.hpp [hunk 4]
```diff
@@ -232,15 +261,31 @@ namespace intercept::types {
             }
             return rcount;
         }
+
+        /**
+         * @brief deletes object
+         *
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:267]:
```suggestion
         * @brief deletes object
```

File: src/client/headers/shared/containers.hpp [hunk 5]
```diff
@@ -232,15 +261,31 @@ namespace intercept::types {
...
+        /**
+         * @brief deletes object
+         * 
+         */
         void destruct() const noexcept {
             delete const_cast<refcount*>(this);
         }
+        /**
+         * @brief wrapper for destruct
+         *
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:274]:
```suggestion
         * @brief wrapper for destruct
```
"wrapper" is not really correct.
This is a customization point so sub classes can run custom code before destruction, which is why its virtual

File: src/client/headers/shared/containers.hpp [hunk 6]
```diff
@@ -232,15 +261,31 @@ namespace intercept::types {
...
         virtual void lastRefDeleted() const { destruct(); }
 
     private:
         virtual int __dummy_refcount_func() const noexcept { return 0; }
     };
+
+
     class game_value_static;
+    /**
+     * @brief Referencing counting class template.
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:285]:
```suggestion
     * @brief Reference counting class template.
```

File: src/client/headers/shared/containers.hpp [hunk 7]
```diff
@@ -253,12 +298,19 @@ namespace intercept::types {
...
+         */
         constexpr ref(Type* other_) noexcept {
             if (other_) other_->add_ref();
             _ref = other_;
         }
-        //Copy from pointer
+
+        /**
+         * @brief Copy from pointer
+         * @param source_ generic type pointer to copy from
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:312]: This is not copy, this is assignment

File: src/client/headers/shared/containers.hpp [hunk 8]
```diff
@@ -267,13 +319,20 @@ namespace intercept::types {
             return *this;
         }
 
-        //Construct from reference
+        /**
+         * @brief Construct from reference
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:323]: This is a copy constructor

File: src/client/headers/shared/containers.hpp [hunk 9]
```diff
@@ -303,12 +369,25 @@ namespace intercept::types {
             return *this;
         }
 
+        /**
+         * @brief Swap references
+         * @param other_ reference to swap from
```
Comment (dedmen) [inline src/client/headers/shared/containers.hpp:374]:
```suggestion
         * @param other_ reference to swap with
```

**Review by dedmen — CHANGES_REQUESTED** [review]

File: src/client/headers/shared/containers.hpp (hunk 5 above)
Comment (killerswin2) [reply src/client/headers/shared/containers.hpp:269]: Ah that's something I must sleep through in my oop class

**Pull request merged**